contrib/dracut/90zfs/mount-zfs.sh.in

```diff
@@ -13,32 +13,69 @@ esac
 # Delay until all required block devices are present.
 udevadm settle
 
-if [ "${root}" = "zfs:AUTO" ] ; then
-	ZFS_DATASET="$(find_bootfs)"
-	if [ $? -ne 0 ] ; then
-		zpool import -N -a ${ZPOOL_IMPORT_OPTS}
-		ZFS_DATASET="$(find_bootfs)"
-		if [ $? -ne 0 ] ; then
-			warn "ZFS: No bootfs attribute found in importable pools."
-			export_all || export_all "-f"
-
-			rootok=0
-			return 1
+ZFS_DATASET="${ZFS_DATASET:-${root#zfs:}}"
+ZFS_POOL="${ZFS_DATASET%%/*}"
+
+# If the initramfs has a cachefile, assume it is alright to use.
+if [ -f /etc/zfs/zpool.cache ]
+then
+	/sbin/zpool import -c /etc/zfs/zpool.cache -aN
+elif /sbin/zpool import -f -N -o readonly=on "${ZFS_POOL}" 2>&1 >/dev/null
+then
+	if ! mount_dataset "${ZFS_DATASET}" ; then
+		/sbin/zpool export -F "${ZFS_POOL}"
+		warn "Failed to mount \"${ZFS_DATASET}\"."
+		rootok=0
+		return 1
+	fi
+
+	# Copy important files to /etc/zfs
+	for i in zpool.cache vdev.conf
```

Comment:
As a note to myself, a `udevadm trigger && udevadm settle` is needed when vdev.conf is read.